includes/class-wc-google-gtag-js.php

```diff
@@ -86,27 +86,86 @@ public function variant_data( $data, $product, $variation ) {
 	}
 
 	/**
-	 * Enqueues JavaScript to build the addImpression event
+	 * Returns Javascript string for Google Analytics events
+	 *
+	 * @param string $event The type of event
+	 * @param array  $data  Event data to be sent
+	 * @return string
+	 */
+	public static function get_event_code( string $event, array $data ): string {
+		return sprintf( "%s('event', '%s', %s)", self::tracker_var(), esc_js( $event ), self::format_event_data( $data ) );
+	}
+
+	/**
+	 * Escape and encode event data
+	 *
+	 * @param array $data Event data to processed and formatted
+	 * @return string
+	 */
+	public static function format_event_data( array $data ): string {
+		$data = apply_filters( 'woocommerce_gtag_event_data', $data );
+
+		// Recursively walk through $data array and escape all values that will be used in JS.
+		array_walk_recursive(
+			$data,
+			function( &$value, $key ) {
+				$value = esc_js( $value );
+			}
+		);
+
+		return wp_json_encode( $data );
+	}
+
+	/**
+	 * Returns an array of category names the product is atttributed to
+	 *
+	 * @param  WC_Product $_product  Product to pull info for
+	 * @return array
+	 */
+	public static function product_get_category_line( $_product ) {
+		$out            = [];
+		$variation_data = $_product->is_type( 'variation' ) ? wc_get_product_variation_attributes( $_product->get_id() ) : false;
+		$categories     = get_the_terms( $_product->get_id(), 'product_cat' );
+
+		if ( is_array( $variation_data ) && ! empty( $variation_data ) ) {
+			$parent_product = wc_get_product( $_product->get_parent_id() );
+			$categories     = get_the_terms( $parent_product->get_id(), 'product_cat' );
+		}
+
+		if ( $categories ) {
+			foreach ( $categories as $category ) {
+				$out[] = $category->name;
+			}
+		}
+
+		return join( '/', $out );
+	}
+
+	/**
+	 * Enqueues JavaScript to build the view_item_list event
 	 *
 	 * @param WC_Product $product
 	 * @param int        $position
 	 */
 	public static function listing_impression( $product, $position ) {
-		if ( isset( $_GET['s'] ) ) {
-			$list = "Search Results";
-		} else {
-			$list = "Product List";
-		}
+		$event_code = self::get_event_code(
+			'view_item_list',
+			array(
+				'items' => array(
+					array(
+						'id'            => $product->get_id(),
+						'name'          => $product->get_title(),
+						'category'      => self::product_get_category_line( $product ),
+						// @codingStandardsIgnoreStart
+						'list'          => isset( $_GET['s'] ) ? __( 'Search Results', 'woocommerce-google-analytics-integration' ) : __( 'Product List', 'woocommerce-google-analytics-integration' ),
+						// @codingStandardsIgnoreEnd
```

Comment:
I wouldn't recommend to turn off all phpcs rules here. We generally also use the format like this:
`// phpcs:ignore Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed`

In my VSCode I have the option "Show sniff source codes in diagnostic messages" enabled which means I can just copy and paste the specific sniff code to ignore.

```suggestion
						// phpcs:ignore WordPress.Security.NonceVerification.Recommended
						'list'          => isset( $_GET['s'] ) ? __( 'Search Results', 'woocommerce-google-analytics-integration' ) : __( 'Product List', 'woocommerce-google-analytics-integration' ),
```

In this case it's important to ignore a specific rule, because we still want it to check that the right i18n domain name is being used. Or any other rules for that line.

Reply:
Thanks for the suggestion Mik - I've updated the [rule being ignored here](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/2b656ac996cdf886583977a39c77a31d7487db39/includes/class-wc-google-gtag-js.php#L158).